proxy: update mcmc and calling conventions

dormando opened the pull request:
upstream fixes: mcmc would return OK to garbage responses, which was
probably causing issues in the past.

This does remove the MCMC_CODE_MISS and replace it with MCMC_CODE_END.

Updated the commit a bit more: status is now either `MCMC_OK` or `MCMC_ERR` and resp.code always has the detail. I feel this is slightly cleaner overall.

I also didn't pull in the tests/* dir from mcmc since that's not really necessary yet.

@smukil - tagging for interest

dormando commented:
starting to lean toward resp.code is the detail and top level function returns 0 or -1.

code could always be positive then? not sure. the parser's a lot cleaner with the single status code; could internally use -CODE for errors and then use that for the rv test and invert it.

dormando commented:
Planning on merging this today or tomorrow.. it should drastically improve bug safety.

smukil commented:
@dormando Just got back from vacation. I'll submit a review on this by tomorrow 3 PM PST if that works!

dormando commented:
@smukil wb! A review would be appreciated even if I merge this ahead of time. See the mcmc repo for more details and the tests.

dormando commented on vendor/mcmc/mcmc.c:
```diff
@@ -173,8 +168,7 @@ static int _mcmc_parse_response(mcmc_ctx_t *ctx, mcmc_resp_t *r) {
                 if (buf[1] == 'N') {
-                    code = MCMC_CODE_MISS;
-                    // TODO: RESP type
+                    code = MCMC_CODE_END;
```
It doesn't; see the changes to the :hit() code. The client code updates align the internal codes with what they are in the actual protocol. Considering it a hit or a miss should be at a higher level.

dormando replied:
previously the return codes were a weird mix of "what the protocol says" vs "what a user thinks it means" and it was leading to bugs

dormando commented:
I think a major thing here is that there are 3-4 different perfectly fine approaches to the calling convention, but whatever it is just needs to be consistent. It can easily be changed if the new approach ends up too hard to work with. So far it seems better and was easier to write tests for.

dormando commented:
I'm also running real slow today so looks like you're winning the race :|

smukil commented on vendor/mcmc/mcmc.h:
```diff
@@ -1,14 +1,16 @@
+#include <sys/uio.h>
+#include <stdint.h>
+
 #define MCMC_OK 0
```
nit: For readability, it might be useful to separate these out with some code comments?
(OK, ERR) -> Status

(NOT_CONN .. HAS_RESULT) -> Connection codes

(CODE_STORED .. CODE_END) -> Successful protocol response codes

(ERR_SHORT .. ERR_VALUE) -> Error protocol response codes

dormando replied:
they already kind of are, as you noted here CODE_* ERR_* etc. I deleted a few others that were unused and violated these principals already. Not sure it's worth splitting the rest out as they're entirely unused at the moment (see the commented out code)

dormando replied:
oh with comments. I guess so?

smukil commented on vendor/mcmc/mcmc.h:
```diff
@@ -1,14 +1,16 @@
+
 #define MCMC_OK 0
 #define MCMC_ERR -1
 #define MCMC_NOT_CONNECTED 1
```
Part of me wonders if it would be useful to separate out the connection codes from the protocol codes as well?
Without knowing the code all that much in depth yet, it might get tricky to track them as part of the same response code as the code grows, even if they're checked at different layers?

dormando replied:
Possibly, but with the read/write functions commented out nothing uses them outside of connect/disconnect. The parser uses WANT_READ but that's just communicated as a normal error code.

If/when the other functions are restored it'll be easier to reason about their return codes.

dormando replied:
Thinking more; potential return codes is kind of a function-by-function (for client facing API calls) thing anyway and if we're documenting things properly it should be at those levels. I think?

smukil replied:
> If/when the other functions are restored it'll be easier to reason about their return codes.
Ack

smukil replied:
> Thinking more; potential return codes is kind of a function-by-function (for client facing API calls) thing anyway and if we're documenting things properly it should be at those levels. I think?

Yea, as long as the contract is in an easily readable format, it may make the codebase easier to navigate. But that's just a nit.

dormando commented:
Think I hear what you're getting at and yes the client is pretty barebones. It needs a doxygen/etc pass for its API. When I first wrote this I thought I'd be returning to it every week and fleshing it out into a full standalone client and I got stuck with just enough time to work on the proxy instead. In this change I removed some of those flags I was fiddling with for as-a-client and added the start of unit tests. As I get time to work on it it'll get some more normal docs, I imagine.

In the meantime it's being used as just a response parser so I haven't been strict. will merge this soon and come back again for more.
